### PR TITLE
oref0-g4-loop

### PR DESCRIPTION
**Pull request by scottleibrand**
Adds:

- oref0-g4-loop (to replace g4-upload and openxshareble)

### CODE REVIEW

**Comment by alimhassam** [comment]
I'll try to review some of this today (cal record stuff at least)

File: lib/determine-basal/determine-basal.js [hunk 1]
```diff
@@ -73,10 +73,13 @@ var determine_basal = function determine_basal(glucose_status, currenttemp, iob_
     }
     if (minAgo > 12 || minAgo < -5) { // Dexcom data is too old, or way in the future
         rT.reason = "If current system time "+systemTime+" is correct, then BG data is too old. The last BG data was read "+minAgo+"m ago at "+bgTime;
-    }
     // if BG is too old/noisy, or is completely unchanging, cancel any high temps and shorten any long zero temps
-    if ( glucose_status.short_avgdelta == 0 && glucose_status.long_avgdelta == 0 ) {
-        rT.reason = "Error: CGM data is unchanged for the past ~45m";
+    } else if ( glucose_status.short_avgdelta == 0 && glucose_status.long_avgdelta == 0 ) {
+        if ( glucose_status.last_cal && glucose_status.last_cal < 3 ) {
```
Comment (alimhassam) [inline lib/determine-basal/determine-basal.js]: probably not a big concern, but does `glucose_status.last_cal < 3`  deal with cgm gaps?

**Comment by rkresha** [comment]
I'd be more than happy to test this out since I use G4 w/ share via BLE (not hardwired) by default.

currently getting this output when trying to connect and pull data. Was on `BlueZ 5.23` and was getting the same error message so I went through the make and install of `BlueZ 5.48` with the same end result

```
root@openaps_echo:/root/myopenaps# cd /root/myopenaps && ps aux | grep -v grep | grep bash | grep -q 'oref0-g4-loop' || oref0-g4-loop 2>&1 | tee -a /var/log/openaps/cgm-loop.log

Starting oref0-g4-loop at Tue May 15 08:55:26 CDT 2018:
unable to initialize libusb: -99
DEXCOM_CGM_ID=SMXXXXXXXX
jq: error: Could not open file cgm/g4-glucose.json: No such file or directory
cgm/g4-glucose.json has < 24h worth of data
rm: cannot remove ‘cgm/g4-glucose.json’: No such file or directory
2018/05/15 08:55:26 PAPERTRAIL environment variable is not set
2018/05/15 08:55:26 read 0 entries from cgm/g4-glucose.json
2018/05/15 08:55:26 BlueZ 5.48: starting discovery
2018/05/15 08:55:27 BlueZ 5.48: discovered DEXCOMRX
2018/05/15 08:55:27 BlueZ 5.48: stopping discovery
2018/05/15 08:55:27 DEXCOMRX: connecting
2018/05/15 08:55:27 DEXCOMRX: pairing
2018/05/15 08:55:28 interface org.bluez.GattCharacteristic1 not found
cat: cgm/g4-glucose.json: No such file or directory
Found  calibration records.
Could not require: /root/myopenaps/cgm/g4-glucose.json { [Error: Cannot find module '/root/myopenaps/cgm/g4-glucose.json'] code: 'MODULE_NOT_FOUND' }
Could not require: /root/myopenaps/cgm/cal.json [SyntaxError: /root/myopenaps/cgm/cal.json: Unexpected end of input]
Could not parse input data:  [TypeError: Cannot call method 'map' of undefined]
jq: error: Could not open file cgm/g4-glucose.json: No such file or directory
date: invalid date ‘@’
-rw-r--r-- 1 root root      0 May 15 00:00 cgm/cgm-glucose.json
-rw-r--r-- 1 root root 109212 May 15 08:54 cgm/glucose.json
Completed oref0-g4-loop at Tue May 15 08:55:29 CDT 2018
```

**Comment by scottleibrand** [comment]
Perhaps @ecc1 knows what would cause that.

**Comment by ecc1** [comment]
Have you done "forget device" on the G4 receiver and restarted Share before doing this? If so, and it still fails, can you pair with it via bluetoothctl?

**Comment by rkresha** [comment]
@ecc1 Yes, I've got two rigs that fight over the pairing with the G4 so I often have to do the forget device option. Can you please provide additional details surrounding what you want me to do after pairing with bluetoothctl?

**Comment by scottleibrand** [comment]
I’ve found that this only works correctly with a single rig configured to pull from each receiver.

**Comment by rkresha** [comment]
While testing I shut the other rig down so I didn't have to deal with the contention.

**Comment by ecc1** [comment]
When you run bluetoothctl, you should see something like this:
```
NEW] Device FA:1C:36:05:CB:C8 DEXCOMRX                                                   
[NEW] Primary Service                                                                     
        /org/bluez/hci0/dev_FA_1C_36_05_CB_C8/service0026                                 
        00001804-0000-1000-8000-00805f9b34fb                                              
        Tx Power                                                                          
[NEW] Characteristic                                                                      
        /org/bluez/hci0/dev_FA_1C_36_05_CB_C8/service0026/char0027                        
        00002a07-0000-1000-8000-00805f9b34fb                                              
        Tx Power Level                                                                    
[NEW] Primary Service                                                                     
        /org/bluez/hci0/dev_FA_1C_36_05_CB_C8/service001d                                 
        0000180a-0000-1000-8000-00805f9b34fb                                              
        Device Information                                                                
[NEW] Characteristic                                                                      
        /org/bluez/hci0/dev_FA_1C_36_05_CB_C8/service001d/char0024                        
        00002a26-0000-1000-8000-00805f9b34fb                                              
        Firmware Revision String                                                          
[NEW] Characteristic                                                                      
        /org/bluez/hci0/dev_FA_1C_36_05_CB_C8/service001d/char0022                        
        00002a27-0000-1000-8000-00805f9b34fb                                              
        Hardware Revision String                                                          
[NEW] Characteristic                                                                      
        /org/bluez/hci0/dev_FA_1C_36_05_CB_C8/service001d/char0020                        
        00002a24-0000-1000-8000-00805f9b34fb                                              
        Model Number String                                                               
[NEW] Characteristic                                                                      
        /org/bluez/hci0/dev_FA_1C_36_05_CB_C8/service001d/char001e                        
        00002a29-0000-1000-8000-00805f9b34fb                                              
        Manufacturer Name String                                                          
[NEW] Primary Service
        /org/bluez/hci0/dev_FA_1C_36_05_CB_C8/service000c
        f0aca0b1-ebfa-f96f-28da-076c35a521db
        Vendor specific
[NEW] Characteristic
        /org/bluez/hci0/dev_FA_1C_36_05_CB_C8/service000c/char001a
        f0acb0cd-ebfa-f96f-28da-076c35a521db
        Vendor specific
...
```
Do you? (Also, is this on an Edison or Raspberry Pi?)

**Comment by ecc1** [comment]
Also, I have only tested with Debian stretch's version of bluez (5.43); I don't compile it from source.

**Comment by rkresha** [comment]
> Do you? (Also, is this on an Edison or Raspberry Pi?)
Yes, I see it detect the devices when it is scanning. This is on an Edison

**Comment by ecc1** [comment]
Does it show all those services and characteristics?  Is there any chance you have both the system version of bluez and your self-compiled one on the system? What do ```which bluetoothctl``` and ```bluetoothctl --version``` report?

**Comment by rkresha** [comment]
@ecc1 
> root@openaps_echo:/root/myopenaps# which bluetoothctl
/usr/local/bin/bluetoothctl
root@openaps_echo:/root/myopenaps# bluetoothctl --version
bluetoothctl: 5.48

**Comment by rkresha** [comment]
I replaced my built blueZ 5.48 with all the locations that

` find / -name bluetoothd` yielded

**Comment by ecc1** [comment]
You didn't answer the question of whether your bluetoothctl is showing the GATT services and characteristics correctly.

I don't know if the issue is with self-compiled bluetoothd, with version 5.48, or with the configuration files. But since others seem to use the (script-driven) self-compiled version, I'm guessing there's some configuration file issue.

The only things I can suggest at this point are: use the bluez package from Debian stretch, or remove all bluetooth config files and rebuild/reinstall your self-compiled version.

**Comment by rkresha** [comment]
@ecc1 
Sorry about that, forgot to include it
> root@openaps_echo:/root/myopenaps# bluetoothctl
[NEW] Controller 43:34:1B:00:XX:XX openaps_echo [default]
[NEW] Device FA:D3:5F:E3:XX:XX DEXCOMRX
[NEW] Primary Service
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service0026
        00001804-0000-1000-8000-00805f9b34fb
        Tx Power
[NEW] Characteristic
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service0026/char0027
        00002a07-0000-1000-8000-00805f9b34fb
        Tx Power Level
[NEW] Primary Service
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service001d
        0000180a-0000-1000-8000-00805f9b34fb
        Device Information
[NEW] Characteristic
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service001d/char0024
        00002a26-0000-1000-8000-00805f9b34fb
        Firmware Revision String
[NEW] Characteristic
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service001d/char0022
        00002a27-0000-1000-8000-00805f9b34fb
        Hardware Revision String
[NEW] Characteristic
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service001d/char0020
        00002a24-0000-1000-8000-00805f9b34fb
        Model Number String
[NEW] Characteristic
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service001d/char001e
        00002a29-0000-1000-8000-00805f9b34fb
        Manufacturer Name String
[NEW] Primary Service
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service000c
        f0aca0b1-ebfa-f96f-28da-076c35a521db
        Vendor specific
[NEW] Characteristic
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service000c/char001a
        f0acb0cd-ebfa-f96f-28da-076c35a521db
        Vendor specific
[NEW] Descriptor
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service000c/char001a/desc001c
        00002902-0000-1000-8000-00805f9b34fb
        Client Characteristic Configuration
[NEW] Characteristic
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service000c/char0018
        f0acb0cc-ebfa-f96f-28da-076c35a521db
        Vendor specific
[NEW] Characteristic
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service000c/char0016
        f0acacac-ebfa-f96f-28da-076c35a521db
        Vendor specific
[NEW] Characteristic
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service000c/char0013
        f0acb20b-ebfa-f96f-28da-076c35a521db
        Vendor specific
[NEW] Descriptor
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service000c/char0013/desc0015
        00002902-0000-1000-8000-00805f9b34fb
        Client Characteristic Configuration
[NEW] Characteristic
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service000c/char0011
        f0acb20a-ebfa-f96f-28da-076c35a521db
        Vendor specific
[NEW] Characteristic
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service000c/char000d
        f0ac2b18-ebfa-f96f-28da-076c35a521db
        Vendor specific
[NEW] Descriptor
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service000c/char000d/desc0010
        00002904-0000-1000-8000-00805f9b34fb
        Characteristic Format
[NEW] Descriptor
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service000c/char000d/desc000f
        00002902-0000-1000-8000-00805f9b34fb
        Client Characteristic Configuration
[NEW] Primary Service
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service0008
        00001801-0000-1000-8000-00805f9b34fb
        Generic Attribute Profile
[NEW] Characteristic
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service0008/char0009
        00002a05-0000-1000-8000-00805f9b34fb
        Service Changed
[NEW] Descriptor
        /org/bluez/hci0/dev_FA_D3_5F_E3_XX_XX/service0008/char0009/desc000b
        00002902-0000-1000-8000-00805f9b34fb
        Client Characteristic Configuration


As for script driven, it was using `BlueZ 5.23` which would've been the script driven version maybe? I was running an older dev branch before which would've been BlueZ 5.46 or 5.47 so the fact it was using 5.23 so something changed it back to an older version.

**Comment by rkresha** [comment]
Trying this
- removing the directories with 5.48
- apt-get remove bluez
- ` find / -name bluetoothd -delete`
- `find / -name "*.pyc" -delete`
- `touch /forcefsck && reboot`
- apt-get install bluez
> root@openaps_echo:~# hciconfig -a
hci0:   Type: BR/EDR  Bus: UART
        BD Address: 43:34:1B:00:1F:AC  ACL MTU: 1021:8  SCO MTU: 64:1
        UP RUNNING PSCAN
        RX bytes:897 acl:0 sco:0 events:49 errors:0
        TX bytes:1059 acl:0 sco:0 commands:40 errors:0
        Features: 0xbf 0xfe 0xcf 0xfe 0xdb 0xff 0x7b 0x87
        Packet type: DM1 DM3 DM5 DH1 DH3 DH5 HV1 HV2 HV3
        Link policy: RSWITCH SNIFF
        Link mode: SLAVE ACCEPT
        Name: 'BlueZ 5.23'
        Class: 0x000110
        Service Classes: Unspecified
        Device Class: Computer, Handheld
        HCI Version: 4.0 (0x6)  Revision: 0xa6
        LMP Version: 4.0 (0x6)  Subversion: 0x410e
        Manufacturer: Broadcom Corporation (15)

> root@openaps_echo:~# bluetoothd -v
5.23

> root@openaps_echo:~# bluetoothctl --version
5.48
```
root@openaps_echo:~# which bluetoothctl
/usr/local/bin/bluetoothctl
root@openaps_echo:~# rm /usr/local/bin/bluetoothctl
root@openaps_echo:~# which bluetoothctl
/usr/bin/bluetoothctl
root@openaps_echo:~# bluetoothctl --version
/usr/local/bin/bluetoothctl: No such file or directory

<LOGOUT of session and LOG BACK IN>

root@openaps_echo:~# which bluetoothctl
/usr/bin/bluetoothctl
root@openaps_echo:~# bluetoothctl --version
5.23
```

Looks better now, somewhat... getting BLE call timeout now
```
root@openaps_echo:/root/myopenaps/cgm# cd /root/myopenaps && ps aux | grep -v grep | grep bash | grep -q 'oref0-g4-loop' || oref0-g4-loop 2>&1 | tee -a /var/log/openaps/cgm-loop.log

Starting oref0-g4-loop at Thu May 17 11:06:42 CDT 2018:
unable to initialize libusb: -99
DEXCOM_CGM_ID=SMXXXXXXXXX
jq: error: Could not open file cgm/g4-glucose.json: No such file or directory
cgm/g4-glucose.json has < 24h worth of data
rm: cannot remove ‘cgm/g4-glucose.json’: No such file or directory
2018/05/17 11:06:43 PAPERTRAIL environment variable is not set
2018/05/17 11:06:43 read 0 entries from cgm/g4-glucose.json
2018/05/17 11:06:43 BlueZ 5.23: removing device DEXCOMRX
2018/05/17 11:06:43 BlueZ 5.23: starting discovery
2018/05/17 11:06:43 BlueZ 5.23: discovered DEXCOMRX
2018/05/17 11:06:43 BlueZ 5.23: stopping discovery
2018/05/17 11:06:43 DEXCOMRX: connecting
2018/05/17 11:06:43 DEXCOMRX: pairing
2018/05/17 11:06:48 BLE call timeout
cat: cgm/g4-glucose.json: No such file or directory
Found  calibration records.
Could not require: /root/myopenaps/cgm/g4-glucose.json { [Error: Cannot find module '/root/myopenaps/cgm/g4-glucose.json'] code: 'MODULE_NOT_FOUND' }
Could not require: /root/myopenaps/cgm/cal.json [SyntaxError: /root/myopenaps/cgm/cal.json: Unexpected end of input]
Could not parse input data:  [TypeError: Cannot call method 'map' of undefined]
jq: error: Could not open file cgm/g4-glucose.json: No such file or directory
date: invalid date ‘@’
```

**Comment by rkresha** [comment]
@ecc1 I misinterpreted your comment about stretch, I was still running jessie. Anyways I'm rebuilding a rig now and trying to get to Stretch. Running into an issue with a missing library on boot for liblz4.1.so was curious for you upgrade path if you used jubilinux 3 image or whether it was jubilinux 2 image and updated via apt-get. Can you please share? Thanks in advance!

**Comment by scottleibrand** [comment]
Any reason not to merge this to 0.7.0-dev so we can cut down on the number of branches we're maintaining and resolving merge conflicts for?

**Comment by rkresha** [comment]
@scottleibrand, this working for Dana via Bluetooth or is she hardwired?

**Comment by danamlewis** [comment]
@rkresha I have one working via bluetooth successfully

**Comment by scottleibrand** [comment]
@rkresha both: we have the receiver plugged into one rig with USB, and paired with bluetooth to another rig.

**Comment by rkresha** [comment]
@danamlewis @scottleibrand ... Jessie? What version of bluetoothd and bluetoothctl ? Edison? I'm trying to understand where I went wrong.

**Comment by rkresha** [comment]
@scottleibrand thanks!

Fresh jubilinux flash and I ran the setup script. 

Current state of the oref0-setup.sh .. syntax error at line 881... missing a `fi` (I can PR if need be) and after setup `oref0-g4-loop` isn't found.

**Comment by scottleibrand** [comment]
That had been already fixed in dev and 0.7.0-dev: I just merged it to go-dexcom as well.  (This is part of the reason I want to go ahead and merge this, so we have fewer branches to manage.)

**Comment by rkresha** [comment]
@scottleibrand so the loop is working now, I had to work through some things `npm install -g` because of the missing oref0-g4-loop, etc which I'm sure are artifacts of the having to keep all the branches up to date. One thing that was odd, was looking at the cgm-loop.log file and it kept reporting using BlueZ 5.23 even though `which` reported it should be picking 5.48. 

Got around it by copying the 5.48 binaries to the following locations
```
cp -p /home/.rootfs/usr/local/bin/bluetoothctl /home/.rootfs/usr/bin/bluetoothctl
cp -p /home/.rootfs/usr/local/bin/bluetoothd /home/.rootfs/usr/lib/bluetooth/bluetoothd
cp -p /home/.rootfs/usr/local/bin/bluetoothd /home/.rootfs/usr/sbin/bluetoothd
```

I shouldn't have to `apt-get remove bluetooth`  correct before making the 5.48, right? 

P.S. thanks for your assistance as I worked my way through

**Comment by scottleibrand** [comment]
Ok.  I don't really know what we should expect from bluez there.  Probably the best path forward is: I merged this PR, then we do further testing by running a from-scratch reflash and reinstall of 0.7.0-dev and run oref0-setup with the g4-go option, and if it doesn't get all the bluez stuff correct, figure out how to fix there.  Any objections to merging this to 0.7.0-dev now?

**Comment by alimhassam** [comment]
+1 for merging. There are a couple other behavior changes here on top of the g4 support which should be merge in order to get a wider audience and testing.

**Pull request merged**